src/metabase/db/util.clj

```diff
@@ -66,3 +68,71 @@
+       ;; possible database conflict due to a concurrent insert. If we want to be more conservative, we would need
+       ;; a per-driver or driver agnostic way to test the exception.
+       ~@body)))
+
+(defn select-or-insert!
+  "Return a database record if it exists, otherwise create it.
+
+   This is more general than using `UPSERT`, `MERGE` or `INSERT .. ON CONFLICT`, and it also allows one to avoid
+   calculating initial values that may be expensive, or require side effects.
+
+   In the case where there is an underlying db constraint to prevent duplicates, this method takes care of handling
+   rejection from the database due to a concurrent insert, and will retry a single time to pick up the existing row.
+   This may result in `insert-fn` being called a second time.
+
+   In the case where there is no underlying db constraint, concurrent calls may still result in duplicates.
+   To prevent this in a database agnostic way, during an existing non-serializable transaction, would be non-trivial."
+  {:style/indent 2}
+  [model select-map insert-fn]
+  (with-conflict-retry
+   (let [select-kvs (mapcat identity select-map)
+         validate   (fn [updated#]
+                      (assert (= select-map (merge select-map (select-keys updated# (keys select-map))))
+                              "This should not be used to change any of the identifying values")
+                      ;; For convenience, we allow the insert-fn to omit fields in the search-map
+                      (merge updated# select-map))]
+     (or (apply t2/select-one model select-kvs)
+         (t2/insert-returning-instance! model (validate (insert-fn)))))))
+
+(defn update-or-insert!
+  "Update a database record, if it exists, otherwise create it.
+
+   This is more general than using `UPSERT`, `MERGE` or `INSERT .. ON CONFLICT`, and it also allows one to avoid
+   calculating initial values that may be expensive, or require side effects.
+
+   In the case where there is an underlying db constraint to prevent duplicates, this method takes care of handling
+   rejection from the database due to a concurrent insert, and will retry a single time to pick up the existing row.
+   This may result in `insert-fn` being called a second time.
+
+   In the case where there is no underlying db constraint, concurrent calls may still result in duplicates.
+   To prevent this in a database agnostic way, during an existing non-serializable transaction, would be non-trivial."
+  {:style/indent 2}
+  [model select-map update-fn]
+  (with-conflict-retry
+   (let [select-kvs (mapcat identity select-map)
+         pks        (t2/primary-keys model)
+         _          (assert (= 1 (count pks)) "This helper does not currently support compound keys")
+         pk-key     (keyword (first pks))
+         validate   (fn [updated]
+                      (assert (= select-map (merge select-map (select-keys updated (keys select-map))))
```

Comment:
This expression is complicated and reusable enough I feel like it deserves it's own name, and since it's duplicated it could be extracted and reduce code length. It could even be in `metabase.util`

Reply:
I didn't think of a good name at the time, and that stopped me extracting it, but I agree it would really help to hide these details here. I'll try find inspiration for a util worthy name, how exciting 😁

Reply:
`maps-conflict?`

Reply:
That's pretty good, clear and concise.

I can't think of many standard functions that have the type in their name, so that's one thing that puts me off.

How about making it slightly more general purpose - a function that returns the `conflicting-keys`.

Reply:
That limits its performance as `maps-conflict?` can short-circuit on the first conflict. but I'm bikeshedding, both are fine for this use case

Reply:
I made it return a lazy sequence, so it still short-circuits.